## Third Party Licenses

This project includes the following third-party tools:

- **Delly**
   - **Description**: Delly is an integrated structural variant (SV) prediction method that can discover, genotype and visualize deletions, tandem duplications, inversions and translocations at single-nucleotide resolution in short-read and long-read massively parallel sequencing data. It uses paired-ends, split-reads and read-depth to sensitively and accurately delineate genomic rearrangements throughout the genome.
   - **Link to License**: [Delly Github](https://github.com/dellytools/delly)
   - **License Text**: The software is licensed under the [BSD 3-Clause License](https://opensource.org/license/BSD-3-Clause)
 
 - **BWA**
   - **Description**: WA is a software package for mapping DNA sequences against a large reference genome, such as the human genome. It consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM.  
   - **Link to License**: [BWA GitHub](https://github.com/lh3/bwa)  
   - **License Text**: The software is licensed under the [GPL-3.0 License](https://www.gnu.org/licenses/gpl-3.0.html).
  
 - **Trimmomatic** 
   - **Description**: Trimmomatic is a flexible read trimming tool for Illumina NGS data.  
   - **Link to License**: [Trimmomatic GitHub](https://github.com/usadellab/Trimmomatic)  
   - **License Text**: The software is licensed under the [GPL-3.0 License](https://www.gnu.org/licenses/gpl-3.0.html).

 - **Picard**
   - **Description**: Picard is a set of Java command line tools for manipulating high-throughput sequencing (HTS) data and formats.  
   - **Link to License**: [Picard GitHub](https://github.com/broadinstitute/picard)  
   - **License Text**: The software is licensed under the [MIT license](https://mit-license.org/).

 - **Samtools** 
   - **Description**: Samtools are tools (written in C using htslib) for manipulating next-generation sequencing data. 
   - **Link to License**: [Samtools GitHub](https://github.com/samtools/samtools)  
   - **License Text**: The software is licensed under the [MIT license](https://mit-license.org/).

 - **Bcftools** 
   - **Description**: BCFtools is a software suite for working with Variant Call Format (VCF) and Binary Variant Call Format (BCF) files.
   - **Link to License**: [Bcftools GitHub](https://github.com/samtools/bcftools)  
   - **License Text**: The software is licensed under the [MIT license](https://mit-license.org/).

 - **GATK**: 
   - **Description**: GATK is a toolkit for variant discovery in high-throughput sequencing data.
   - **Link to License**: [GATK GitHub](https://github.com/broadinstitute/gatk)  
   - **License Text**: The software is licensed under the [Apache 2.0 License](https://www.apache.org/licenses/LICENSE-2.0).

 - **CNVKit**: 
   - **Description**: CNVkit is a command-line toolkit and Python library for detecting copy number variants and alterations genome-wide from high-throughput sequencing.
   - **Link to License**: [CNVkit GitHub](https://github.com/etal/cnvkit)  
   - **License Text**: The software is licensed under the [Apache 2.0 License](https://www.apache.org/licenses/LICENSE-2.0).

 - **Nextflow**: 
   - **Description**: Nextflow is a workflow system for creating scalable, portable, and reproducible workflows. It is based on the dataflow programming model, which greatly simplifies the writing of parallel and distributed pipelines, allowing you to focus on the flow of data and computation. Nextflow can deploy workflows on a variety of execution platforms, including your local machine, HPC schedulers, AWS Batch, Azure Batch, Google Cloud Batch, and Kubernetes. Additionally, it supports many ways to manage your software dependencies, including Conda, Spack, Docker, Podman, Singularity, and more.
   - **Link to License**: [Nextflow GitHub](https://github.com/nextflow-io/nextflow)  
   - **License Text**: The software is licensed under the [Apache 2.0 License](https://www.apache.org/licenses/LICENSE-2.0).

